[PC-1257] IMU Working Example Update (#1064)
arduino/docs-content: content/hardware/06.nicla/boards/nicla-voice/tutorials/user-manual/content.md

- **Microcontroller**: at the heart of the Nicla Voice is the nRF52832, a powerful and versatile System-on-Chip (SoC) from Nordic® Semiconductor. The nRF52832 is built around a 32-bit Arm® Cortex®-M4 processor running at 64 MHz.
- **Speech and sensor-fusion processor**: the board features the NDP120 Neural Decision Processor™, an ultra-low power always-on audio and sensor-fusion processor from Syntiant®, which enables several applications, including echo-cancellation, beamforming, noise suppression, speech enhancement, speaker identification, keyword spotting, acoustic event and scene classification, and multi-sensor fusion. 

> - **Onboard advanced motion sensors**: the board features the BMI270, a high-precision Inertial Measurement Unit (IMU) by Bosch® Sensortec, which combines a 3-axis accelerometer and a 3-axis gyroscope for precise motion tracking and orientation detection. The board also features the BMM150, a compact geomagnetic sensor from Bosch® Sensortec with a 3-axis magnetometer.

Reviewer: Just a small rephrase: 
- **Onboard advanced motion sensors**: the board features the BMI270, a high-precision Inertial Measurement Unit (IMU) from Bosch® Sensortec, which combines a 3-axis accelerometer and a 3-axis gyroscope for precise motion tracking and orientation detection. The board also features the BMM150, a compact geomagnetic sensor from Bosch® Sensortec with a 3-axis magnetometer.

Author: Suggestion added!

![Nicla Voice onboard magnetometer](assets/user-manual-12.png)

#### Accelerometer and Gyroscope Data

> The example code below shows how to get acceleration (in meters per second squared) and angular velocity (in degrees per second) data from the onboard IMU and streams it to the Arduino IDE Serial Monitor.

Reviewer 2: ```suggestion
The example code below shows how to get acceleration (m/s) and angular velocity (in degrees per second) data from the onboard IMU and streams it to the Arduino IDE Serial Monitor.
```

  Serial.print("x_acc:");
  Serial.print(x_acc);
  Serial.print(",");
  Serial.print("y_acc:");
  Serial.print(y_acc);
  Serial.print(",");
  Serial.print("z_acc:");
  Serial.println(z_acc);

  // Print gyroscope data (expressed in degrees per second). 
  Serial.print("x_gyr:");
  Serial.print(x_gyr);
  Serial.print(",");
  Serial.print("y_gyr:");
  Serial.print(y_gyr);
  Serial.print(",");
  Serial.print("z_gyr:");
  Serial.println(z_gyr);

  delay(1000);
}
```

Here you can find a step-by-step explanation of the code:

First, the necessary libraries are included: 

- `NDP.h` and `BMI270_Init.h` for the Nicla Voice board's basic functions and the IMU control.
- Macros are defined for checking the status of the IMU; these macros allow the sketch to detect and handle sensor errors. 

Next, user functions `ledBlueOn()`, `ledGreenOn()`, and `ledRedBlink()` definition: 

- These functions allow the onboard LEDs to flash specific colors to indicate different states: blue for a successful match, green for an event, and red to indicate an error.

Next, in the `setup()` function:

- The serial communication is initialized at a baud rate of 115200.

> - The Nicla Voice board is initialized, and the LDO is disabled.

Reviewer 2: explain in a couple of words what is the LDO and what we want to disable it

Something like:

The nicla board is initialized, and the LDO (the power regulator responsible of putting the Nicla on power-saving modes) is disabled to avoid reading disconnections for the IMU.

- Error and event handlers are initialized.
- NDP processor is initialized; this process includes populating the external Flash memory of the board with the NDP processor internal microcontroller firmware (`mcu_fw_120_v91.synpkg`), the NDP processor internal DSP firmware (`dsp_firmware_v91.synpkg`), and the ML model (`ei_model.synpkg`). 
- The BMI270 sensor is initialized; this includes a software reset, loading the sensor configuration, and setting it into normal power mode with the accelerometer and gyroscope operational. 

Finally, in the `loop()` function:

- Memory is allocated for the sensor data; data is then read from the sensor and stored in this allocated space.

> - Raw sensor data is then parsed and extracted into raw accelerometer and gyroscope data. This data is represented as 16-bit signed integers ranging from -32,768 to 32,767.

Reviewer 2: I have some fears here:

32,767 it s 32767 (32 thousand bla bla bla)

Should we put the USA comma here (European point) @jhansson-ard or without it to avoid confusion across different users in different parts of the world?

Reviewer 3: 1. 32,767 (thirty-two thousand seven hundred and seven)
2. 32.676 (thirty-two point six seven six)
3. 32 767 (thirty-two thousand seven hundred and seven)

Either we decide to go with 1 or 3. I can't find any examples in the documentation to us leaning towards either. Which of these do you prefer?

Author: Should we go now with 1 or 3 @jcarolinares? I like 3 more than 1, but 1 can potentially have more users (USA).

Reviewer 3: Added this to the Style Guide, thanks guys!

- Raw sensor data is converted into understandable and standard unit measurements; for the accelerometer, data is converted to meters per second squared, and for the gyroscope, data is converted to degrees per second. 
- Converted accelerometer and gyroscope data are printed to the Serial Monitor, allowing one to observe sensor data in real time.

After uploading the example code, you should see accelerometer and gyroscope data on the IDE's Serial Monitor as shown below:

![Nicla Voice onboard IMU data on the IDE's Serial Monitor](assets/user-manual-13.png)

> Let's use also the Arduino IDE Serial Plotter to test the example IMU sketch; let's visualize first only accelerometer data. To do this, **comment the gyroscope data output to the IDE's Serial monitor and upload the example sketch again**; then, open the IDE's Serial Plotter by navigating to **Tools -> Serial Plotter**. After a while, you should see a real-time graph showing data from the board's onboard accelerometer, as shown below (move the board):

Reviewer 2: As a user I don't understand what and what I have to comment.

I suggest to improve the phrase and also add a small block code below with that part of the code where the user can be the part that is commented and the part that is uncommented

Author: You are right, I added the code section and changed a little bit the phrase.

> Let's use also the Arduino IDE Serial Plotter to test the example IMU sketch; let's visualize first only accelerometer data. To do this, **comment the gyroscope data output to the IDE's Serial monitor and upload the example sketch again**; then, open the IDE's Serial Plotter by navigating to **Tools -> Serial Plotter**. After a while, you should see a real-time graph showing data from the board's onboard accelerometer, as shown below (move the board):

Reviewer 3: ```suggestion
Let's use also the Arduino IDE Serial Plotter to test the example IMU sketch; let's visualize first only accelerometer data. To do this, **comment the gyroscope data output to the IDE's Serial monitor and upload the example sketch again**; then, open the IDE's Serial Plotter by navigating to **Tools > Serial Plotter**. After a while, you should see a real-time graph showing data from the board's onboard accelerometer, as shown below (move the board):

```

![Nicla Voice onboard accelerometer data on the IDE's Serial Plotter](assets/user-manual-14.gif)

When the board is not moving, you should see acceleration measurements close to zero on the x and y-axis, while the z-axis will be close to 1g (approximately 9.81 m/s^2). If you want to visualize gyroscope readings, uncomment the gyroscope data output and comment on the accelerometer data output; when the board is not moving, you should see gyroscope measurements on the three-axis close to zero.

#### IMU and Machine Learning

> The example code below demonstrates using the Nicla Voice board to perform Machine Learning inference on IMU data. The code sets up event indicators using the onboard RGB LED and sends IMU data to the NDP processor for inference. The example can be found in the board's built-in examples by navigating to **File -> Examples -> NDP -> IMUDemo**.

Reviewer 3: ```suggestion
The example code below demonstrates using the Nicla Voice board to perform Machine Learning inference on IMU data. The code sets up event indicators using the onboard RGB LED and sends IMU data to the NDP processor for inference. The example can be found in the board's built-in examples by navigating to **File > Examples > NDP > IMUDemo**.

```

- Error and event handlers are initialized.
- NDP processor is initialized; this process includes populating the external Flash memory of the board with the NDP processor internal microcontroller firmware (`mcu_fw_120_v91.synpkg`), the NDP processor internal DSP firmware (`dsp_firmware_v91.synpkg`), and the ML model (`ei_model.synpkg`). 
- The BMI270 sensor is initialized; this includes a software reset, loading the sensor configuration, and setting it into normal power mode with the accelerometer and gyroscope operational. 

Finally, in the `loop()` function:

- Memory is allocated for the sensor data; data is then read from the sensor and stored in this allocated space.
- Raw sensor data is then parsed and extracted into raw accelerometer and gyroscope data. This data is represented as 16-bit signed integers ranging from -32 768 to 32 767.
- Raw sensor data is converted into understandable and standard unit measurements; for the accelerometer, data is converted to meters per second squared, and for the gyroscope, data is converted to degrees per second. 

> - Converted accelerometer and gyroscope data are printed to the Serial Monitor, allowing one to observe sensor data in real time.

Reviewer: Just a small rephrase: 
- Converted accelerometer and gyroscope data are printed on the Serial Monitor, allowing the user to observe sensor data in real-time.

Author: Suggestion added!

After uploading the example code, you should see accelerometer and gyroscope data on the IDE's Serial Monitor as shown below:

![Nicla Voice onboard IMU data on the IDE's Serial Monitor](assets/user-manual-13.png)

> Let's use also the Arduino IDE Serial Plotter to test the example IMU sketch; let's visualize first only accelerometer data. To do this, comment the gyroscope data output as shown below: 

Reviewer: Just a small rephrase: 
Let's use also the Arduino IDE Serial Plotter to test the example IMU sketch; let's start visualizing only accelerometer data. To do so, comment the gyroscope data output as shown below:

Author: Suggestion added!

  Serial.print(",");
  Serial.print("y_acc:");
  Serial.print(y_acc);
  Serial.print(",");
  Serial.print("z_acc:");
  Serial.println(z_acc);

  /* Print gyroscope data (expressed in degrees per second). 
  Serial.print("x_gyr:");
  Serial.print(x_gyr);
  Serial.print(",");
  Serial.print("y_gyr:");
  Serial.print(y_gyr);
  Serial.print(",");
  Serial.print("z_gyr:");
  Serial.println(z_gyr); */
```

Upload the example sketch again and open the IDE's Serial Plotter by navigating to **Tools > Serial Plotter**. After a while, you should see a real-time graph showing data from the board's onboard accelerometer, as shown below (move the board):

![Nicla Voice onboard accelerometer data on the IDE's Serial Plotter](assets/user-manual-14.gif)

When the board is not moving, you should see acceleration measurements close to zero on the x and y-axis, while the z-axis will be close to 1g (approximately 9.81 m/s<sup>2</sup>). If you want to visualize gyroscope readings, uncomment the gyroscope data output and comment on the accelerometer data output; when the board is not moving, you should see gyroscope measurements on the three-axis close to zero.

#### IMU and Machine Learning

> The example code below demonstrates using the Nicla Voice board to perform Machine Learning inference on IMU data. The code sets up event indicators using the onboard RGB LED and sends IMU data to the NDP processor for inference. The example can be found in the board's built-in examples by navigating to **File > Examples > NDP > IMUDemo**.

Reviewer: Just a small rephrase: 
The example code below demonstrates how to use the Nicla Voice board to perform Machine Learning inference on IMU data. The code sets up event indicators using the onboard RGB LED and sends IMU data to the NDP processor for inference. The example can be found in the board's built-in examples by navigating to **File > Examples > NDP > IMUDemo**.

Author: Suggestion added!
